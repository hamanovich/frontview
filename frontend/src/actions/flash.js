import { ADD_FLASH_MESSAGE, DELETE_FLASH_MESSAGE, DELETE_FLASH_MESSAGES } from './types';

export const addFlashMessage = payload => ({
  type: ADD_FLASH_MESSAGE,
  payload,
});

export const deleteFlashMessage = id => ({
  type: DELETE_FLASH_MESSAGE,
  id,
});

export const deleteFlashMessages = () => ({
  type: DELETE_FLASH_MESSAGES,
});
