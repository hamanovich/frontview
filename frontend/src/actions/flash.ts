import { FlashMessageType } from '../propTypes';
import {
  ADD_FLASH_MESSAGE,
  DELETE_FLASH_MESSAGE,
  DELETE_FLASH_MESSAGES,
} from './types';

interface addFlashMessageAction {
  type: typeof ADD_FLASH_MESSAGE;
  payload: FlashMessageType;
}

interface deleteFlashMessageAction {
  type: typeof DELETE_FLASH_MESSAGE;
  id: string;
}

interface deleteFlashMessagesAction {
  type: typeof DELETE_FLASH_MESSAGES;
}

export const addFlashMessage = (payload: FlashMessageType) => ({
  type: ADD_FLASH_MESSAGE,
  payload,
});

export const deleteFlashMessage = (id: string | undefined) => ({
  type: DELETE_FLASH_MESSAGE,
  id,
});

export const deleteFlashMessages = () => ({
  type: DELETE_FLASH_MESSAGES,
});

export type FlashActionTypes =
  | addFlashMessageAction
  | deleteFlashMessageAction
  | deleteFlashMessagesAction;
