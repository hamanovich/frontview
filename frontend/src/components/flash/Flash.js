import React from 'react';
import { func, shape, string } from 'prop-types';
import FontAwesome from 'react-fontawesome';
import MarkdownRenderer from 'react-markdown-renderer';
import styled from 'styled-components';
import classNames from 'classnames';

import compose from 'recompose/compose';
import lifecycle from 'recompose/lifecycle';

import Alert from 'react-bootstrap/Alert';
import Button from 'react-bootstrap/Button';

const enhance = compose(
  lifecycle({
    componentDidMount() {
      setTimeout(this.props.close, 4000);
    },
  }),
);

const FlashItem = styled(Alert)`
  display: flex;
  justify-content: space-between;
  position: fixed;
  z-index: 10;
  left: 50%;
  transform: translateX(-50%);
  top: 52px;
  align-items: center;
  min-width: 290px;
  max-width: 90%;

  button {
    margin-left: 1.5rem;
  }

  + .alert {
    margin-top: 4rem;

    + .alert {
      margin-top: 12rem;
    }
  }
`;

export const Flash = ({ close, message }) => (
  <FlashItem
    variant={classNames({
      danger: message.type === 'error',
      success: message.type === 'success',
      warning: message.type === 'warn',
    })}>
    <MarkdownRenderer markdown={message.text} />
    <Button onClick={close} className="close">
      <FontAwesome name="times" />
    </Button>
  </FlashItem>
);

Flash.propTypes = {
  message: shape({
    type: string,
    text: string,
  }).isRequired,
  close: func.isRequired,
};

export default enhance(Flash);
