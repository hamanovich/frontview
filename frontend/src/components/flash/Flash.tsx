import React, { FC, ReactNode } from 'react';
import FontAwesome from 'react-fontawesome';
import MarkdownRenderer from 'react-markdown-renderer';

import compose from 'recompose/compose';
import lifecycle from 'recompose/lifecycle';

import Button from 'react-bootstrap/Button';

import { FlashProps } from './models';

import { FlashItem } from './style';

const enhance = compose<FlashProps, FlashProps>(
  lifecycle<FlashProps, {}>({
    componentDidMount() {
      setTimeout(this.props.close, 4000);
    },
  }),
);

export const Flash: FC<FlashProps> = ({ close, message }) => {
  const variant =
    message.type === 'error'
      ? 'danger'
      : message.type === 'success'
      ? 'success'
      : 'warning';
  return (
    <FlashItem variant={variant}>
      <MarkdownRenderer markdown={message.text} />
      <Button onClick={close} className="close">
        <FontAwesome name="times" />
      </Button>
    </FlashItem>
  );
};

export default enhance(Flash);
