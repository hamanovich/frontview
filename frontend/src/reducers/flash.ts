import shortid from 'shortid';

import { FlashMessageType } from '../propTypes';

import {
  ADD_FLASH_MESSAGE,
  DELETE_FLASH_MESSAGE,
  DELETE_FLASH_MESSAGES,
} from '../actions/types';
import { FlashActionTypes } from '../actions/flash';

const initialState: FlashMessageType[] = [];

export default (
  state = initialState,
  action: FlashActionTypes,
): FlashMessageType[] => {
  switch (action.type) {
    case ADD_FLASH_MESSAGE: {
      const { type, text } = action.payload;

      return [...state, { id: shortid.generate(), type, text }];
    }

    case DELETE_FLASH_MESSAGE: {
      const index: number = state.findIndex(
        (msg: FlashMessageType) => msg.id === action.id,
      );

      if (index >= 0) {
        return [...state.slice(0, index), ...state.slice(index + 1)];
      }

      return state;
    }

    case DELETE_FLASH_MESSAGES:
      return initialState;

    default:
      return state;
  }
};
