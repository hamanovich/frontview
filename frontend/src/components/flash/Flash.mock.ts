import { action } from '@storybook/addon-actions';
import { FlashProps } from './Flash';

export const mockDataSuccess: FlashProps = {
  close: action('close'),
  message: {
    id: '1',
    type: 'success',
    text: 'Well Done!',
  },
};

export const mockDataWithMarkdown: FlashProps = {
  ...mockDataSuccess,
  message: {
    ...mockDataSuccess.message,
    id: '4',
    text: 'The `Question` with id=__123__ was removed. _Good job_!',
  },
};

export const mockDataWarning: FlashProps = {
  ...mockDataSuccess,
  message: {
    id: '2',
    type: 'warning',
    text: 'Watch out!',
  },
};

export const mockDataDanger: FlashProps = {
  ...mockDataSuccess,
  message: {
    id: '3',
    type: 'error',
    text: 'Be careful!',
  },
};
