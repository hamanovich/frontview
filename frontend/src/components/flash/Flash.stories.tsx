import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, object } from '@storybook/addon-knobs';

import { Flash } from './Flash';
import {
  mockDataSuccess,
  mockDataWarning,
  mockDataDanger,
  mockDataWithMarkdown,
} from './Flash.mock';

storiesOf('Flash|Flash Item', module)
  .addDecorator(withKnobs)
  .add('Success', () => (
    <Flash
      {...mockDataSuccess}
      message={object('message', { ...mockDataSuccess.message })}
    />
  ))
  .add('Warning', () => <Flash {...mockDataWarning} />)
  .add('Danger', () => <Flash {...mockDataDanger} />)
  .add('With Markdown', () => <Flash {...mockDataWithMarkdown} />);
