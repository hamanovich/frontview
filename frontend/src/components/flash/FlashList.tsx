import React, { FunctionComponent, Fragment } from 'react';
import { connect } from 'react-redux';

import Flash from './Flash';
import { deleteFlashMessage } from '../../actions/flash';

import { FlashMessageType } from '../../propTypes';

type FlashListProps = {
  messages: FlashMessageType[];
  deleteFlashMessage: (id: string | undefined) => void;
};

export const FlashList: FunctionComponent<FlashListProps> = ({
  messages,
  deleteFlashMessage,
}) => (
  <Fragment>
    {messages.map((message: FlashMessageType) => (
      <Flash
        key={message.id}
        message={message}
        close={() => deleteFlashMessage(message.id)}
      />
    ))}
  </Fragment>
);

const mapStateToProps = (state: { flash: FlashMessageType[] }) => ({
  messages: state.flash,
});

export default connect(
  mapStateToProps,
  { deleteFlashMessage },
)(FlashList);
