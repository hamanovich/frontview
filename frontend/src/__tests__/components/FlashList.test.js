import React from 'react';
import { shallow } from 'enzyme';

import { FlashList } from '../../components/flash/FlashList';

describe('<Flash/>', () => {
  const props = {
    messages: [
      {
        id: 'ABC',
        type: 'error',
        text: 'Very negative text',
      },
    ],
    deleteFlashMessage: jest.fn(),
  };

  const flashList = shallow(<FlashList {...props} />);

  it('renders <Flash /> component', () => {
    expect(flashList).toMatchSnapshot();
  });

  it('updates props correctly', () => {
    expect(flashList.find('Fragment').children()).toHaveLength(1);

    flashList.setProps({
      messages: [...props.messages, { id: 'DEF', type: 'success', text: 'Very positive text' }],
    });

    expect(flashList.find('Fragment').children()).toHaveLength(2);

    expect(flashList).toMatchSnapshot();
  });

  describe('when user wants to close the Flash message', () => {
    beforeEach(() => {
      flashList
        .find('Flash')
        .at(0)
        .props()
        .close();
    });

    it('invokes `deleteFlashMessage` callback with id', () => {
      expect(props.deleteFlashMessage).toHaveBeenCalledWith(props.messages[0].id);
    });
  });
});
