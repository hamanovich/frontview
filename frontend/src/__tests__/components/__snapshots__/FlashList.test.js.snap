// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FlashList /> renders <FlashList /> component 1`] = `
<Fragment>
  <Memo(Flash)
    close={[Function]}
    key="ABC"
    message={
      Object {
        "id": "ABC",
        "text": "Very negative text",
        "type": "error",
      }
    }
  />
</Fragment>
`;

exports[`<FlashList /> updates props correctly 1`] = `
<Fragment>
  <Memo(Flash)
    close={[Function]}
    key="ABC"
    message={
      Object {
        "id": "ABC",
        "text": "Very negative text",
        "type": "error",
      }
    }
  />
  <Memo(Flash)
    close={[Function]}
    key="DEF"
    message={
      Object {
        "id": "DEF",
        "text": "Very positive text",
        "type": "success",
      }
    }
  />
</Fragment>
`;
